### PR TITLE
Bug-fix CSV Agent - Parsing Tool Inputs

### PR DESCRIPTION
**Pull request by juandiegobalbi**
Issue & Description:
Suggested changes are due to encountering constant constant errors when using the Pandas DataFrame Agent due to parsing the python code that comes in the responses from the LLMs.
To address the issue I propose json serialization of the arguments of the function call at the OpenAIFunctionsAgent, as well as a more detailed ArgSchema for the PythonAstREPLTool, and finally an addition to the input query sanitize function, since i've noticed that more often than not, OpenAI's GPT-3.5-turbo make function calls but instead of sending the "query" param as a plan python code string, it tries to serialize it into some sort of "{'query': 'python code snippet'}" string, which is just not what we want.

### CODE REVIEW

File: libs/langchain/langchain/agents/openai_functions_agent/base.py [hunk 1]
```diff
@@ -109,7 +109,9 @@ def _parse_ai_message(message: BaseMessage) -> Union[AgentAction, AgentFinish]:
     if function_call:
         function_name = function_call["name"]
         try:
-            _tool_input = json.loads(function_call["arguments"])
+            # escape special characters
+            _tool_input = json.dumps(function_call["arguments"])
+            _tool_input = json.loads(_tool_input)
```
Comment (eyurtsev) [inline libs/langchain/langchain/agents/openai_functions_agent/base.py:114]: could you check if this works for you?

```python
_tool_input = json.loads(function_call["arguments"], strict=False)
```

**Review by eyurtsev — CHANGES_REQUESTED** [review]
@juandiegobalbi could you add unit-tests? And can you try implementing the suggestion in the code to see if it'll work for you? (without the dumps step)

**Comment by eyurtsev** [comment]
@juandiegobalbi Thanks for the contribution! There might not be existing unit tests for the parsing logic. If you're able to provide an example of a message on which the parsing fails, we can incorporate the fix into the code and add an appropriate unit test :)

**Comment by juandiegobalbi** [comment]
Was thinking the exact same thing haha

Yeah take look:

langchain.schema.output_parser.OutputParserException: Could not parse tool
input: {'name': 'python', 'arguments': "import pandas as pd\nimport
matplotlib.pyplot as plt\n\n# Calculate total sales by
month\ndf['ORDERDATE'] = pd.to_datetime(df['ORDERDATE'])\ndf['MONTH_YEAR']
= df['ORDERDATE'].dt.strftime('%b-%Y')\ntotal_sales_by_month =
df.groupby('MONTH_YEAR')['SALES'].sum().reset_index()\n\n# Sort by
month-year\ntotal_sales_by_month['MONTH_YEAR'] =
pd.to_datetime(total_sales_by_month['MONTH_YEAR'],
format='%b-%Y')\ntotal_sales_by_month =
total_sales_by_month.sort_values('MONTH_YEAR')\n\n# Create Combo
Chart\nfig, ax1 = plt.subplots(figsize=(12, 6))\n\n# Bar
Chart\nax1.bar(total_sales_by_month['MONTH_YEAR'],
total_sales_by_month['SALES'],
color='tab:blue')\nax1.set_xlabel('Month/Year')\nax1.set_ylabel('Total
Sales')\nax1.set_title('Monthly Sales Comparison')\n\n# Line Chart\nax2 =
ax1.twinx()\nax2.plot(total_sales_by_month['MONTH_YEAR'],
total_sales_by_month['SALES'], color='tab:red',
marker='o')\nax2.set_ylabel('Total Sales')\n\n# Rotate x-axis
labels\nplt.xticks(rotation=45)\n\n# Save the
figure\nplt.savefig('monthly_sales_comparison.png')\n\n# Show the
figure\nplt.show()"} because the `arguments` is not valid JSON.

And this happens very often when the message that is trying to pass refers
to python code in this case.

File: libs/langchain/langchain/agents/openai_functions_agent/base.py (hunk 1 above)
Comment (juandiegobalbi) [reply libs/langchain/langchain/agents/openai_functions_agent/base.py:114]: @eyurtsev this approach doesn't fix the issue btw

**Review by hwchase17 — COMMENTED** [review]
im all for the better docstring on the tool

im reasonably for the new regex, although lets add some tests

I don't understand/love the changes to openai functions agent

do you have any test runs (on langsmith) where this fails?

**Comment by juandiegobalbi** [comment]
Hey @hwchase17,
After some intense debugging, I believe i found the underlying issue.
Unfortunately i do not have access to LangSmith Hub, so i cannot share my runs, however i'll try to share my findings here (unless you prefer a better approach :))

Im initializing a csv_agent/pandas_df_agent of type OPENAI_FUNCTIONS, using gpt-3.5-turbo.

If you pay attention to the error message I shared before, i noticed something interesting (i'll just share the first few words that give away the issue):

```
langchain.schema.output_parser.OutputParserException: Could not parse tool
input: {'name': 'python', 'arguments': "import pandas as pd\nimport
```

The LLM is trying to invoke a 'python' function which was never really a part of the "functions" parameter in the first place (pandas agent only gets PythonAstREPLTool on the tool set - agents/agent_toolkits/pandas/base.py - which is named 'python_repl_ast').

This means that the model is hallucinating a function that is not even implemented.
I found that some users already experienced this same behaviour (https://community.openai.com/t/function-calling-a-python-function-is-frequently-called-even-though-it-does-not-exist-in-the-functions-parameter/264481).


I could fix this behaviour through some prompt engineering in the Pandas Agent Prefix + a modification to the python_repl_ast tool name (from my tests, it appears that everytime 'python' is present the tool name the model invokes the non-existent 'python' tool).

By the way, the changes proposed under the openai_functions_agent would be reverted, as well as the new regex on the sanitization function. These were just fixing a problem that didnt quite reside there.

In conclusion, i'd be happy to push my final changes of the solution i described, but i'd like to know beforehand how reluctant to changing the repl tool's name are you, since my solution depends on it.

Cheers!

**Comment by efriis** [comment]
Sorry we never circled back on this one - we probably won't rename the whole tool for this, and it's a different enough change that it might make sense as a different PR if you feel it's still valuable.

Because renaming the AST tool is unlikely, I'll close this PR for now, and let me know if you'd like it reopened!
